server/prisma-rs/libs/datamodel/src/dml/validator/mod.rs

```diff
@@ -98,13 +127,213 @@ impl Validator {
+        let mut relation_has_embedding: HashMap<(String, String, Option<String>), (bool, dml::FieldArity)> =
+            HashMap::new();
+
+        for model in schema.models() {
+            for field in model.fields() {
+                if let dml::FieldType::Relation(rel) = &field.field_type {
+                    // Remember if we have an explicit embedding and our arity.
+                    relation_has_embedding.insert(
+                        (model.name.clone(), rel.to.clone(), rel.name.clone()),
+                        (rel.to_fields.len() > 0, field.arity),
+                    );
+                }
+            }
+        }
+
+        // Iterate and mutate models.
+        for model in schema.models_mut() {
+            let model_name = model.name.clone();
+            for field in model.fields_mut() {
+                if let dml::FieldType::Relation(rel) = &mut field.field_type {
+                    // Do we have an embedding, or does our neighbor have an embedding?
+                    let (relation_has_embedding, related_exists, related_is_list) =
+                        match relation_has_embedding.get(&(rel.to.clone(), model_name.clone(), rel.name.clone())) {
+                            Some((has, dml::FieldArity::List)) => (*has, true, true),
+                            Some((has, _)) => (*has, true, false),
+                            None => (false, false, false),
+                        };
+
+                    let we_have_embedding = rel.to_fields.len() > 0;
+                    let we_are_list = field.arity == dml::FieldArity::List;
+
+                    // Set to_fields to ID if:
+                    // * Embedding is not already set and we are not a list.
+                    // * Our related side does exist and has no embedding and our name is smaller
+                    // * Our related side does not exist.
+                    // * Our related side is a list.
+                    if !we_have_embedding
+                        && !we_are_list
+                        && (!related_exists || related_is_list || (!relation_has_embedding && (model_name < rel.to)))
+                    {
+                        rel.to_fields = id_per_model.get(&rel.to).expect(STATE_ERROR).clone();
+                    }
+                }
+            }
+        }
+    }
+
+    /// Identifies and adds missing back relations.
+    fn add_missing_back_relations(&self, schema: &mut dml::Schema) {
+        let mut missing_back_relations = Vec::new();
+
+        for model in schema.models() {
+            missing_back_relations.append(&mut self.find_fields_with_missing_back_relation(model, schema));
+        }
+
+        for (forward, backward) in missing_back_relations {
+            let model = schema.find_model_mut(&forward.to).expect(STATE_ERROR);
+
+            // TODO: Ugly generated name for now.
+            model.add_field(dml::Field::new_generated(
+                &format!("generated_{}", backward.to),
```

Comment:
I would expect the name of the generated back relation field to not be important. Anyway, I'm very open to changing that.

Reply:
It is actually important. That field will also be visible in the client. We want to user the camelCased variant of the name of the related type in this case.